cmd/metal-api/internal/metal/provisioning.go

```diff
@@ -100,42 +134,40 @@ func (p provisioningEventSequence) containsEvent(event ProvisioningEventType) bo
 	return result
 }
 
-func (e *ProvisioningEvent) hasExpectedSuccessor(log *zap.SugaredLogger, actualSuccessor ProvisioningEventType) bool {
-	currentEvent := e.Event
-
-	expectedSuccessors, ok := expectedSuccessorEventMap[currentEvent]
-	if !ok {
-		log.Errorw("successor map does not contain an expected successor for event", "event", currentEvent)
-		return false
+func postEvent(currentFSM *fsm.FSM, event ProvisioningEventType) (*fsm.FSM, error) {
+	var nextFSM *fsm.FSM
+	err := currentFSM.Event(string(event))
+	if err != nil && err.Error() != "no transition" {
+		nextFSM = newProvisioningFSM()
+		_ = nextFSM.Event(string(event)) // Should not give an error
+	} else {
+		nextFSM = currentFSM
 	}
+	return nextFSM, err
+}
 
-	return expectedSuccessors.containsEvent(actualSuccessor)
+func newProvisioningFSM() *fsm.FSM {
+	return fsm.NewFSM(
+		FsmStateInitial,
+		events,
+		callbacks,
+	)
 }
 
 // CalculateIncompleteCycles calculates the number of events that occurred out of order. Can be used to determine if a machine is in an error loop.
 func (p *ProvisioningEventContainer) CalculateIncompleteCycles(log *zap.SugaredLogger) string {
+	fsm := newProvisioningFSM()
 	incompleteCycles := 0
-	atLeastOneTimeCompleted := true
-	var successor ProvisioningEventType
-	for i, event := range p.Events {
-		if successor != "" && !event.hasExpectedSuccessor(log, successor) {
+	var err error
```

Comment:
I don't think you need to take care so much of the "incomplete cycles"-part for now because it's just sitting on top of the rest of the functionality. We can probably discuss to replace it by a single "CrashLoop"-flag or something in a later iteration.

So I think, I would like if this function is just kept as it is and we clean it up in the end when we don't need it anymore.